File: multirateSdc/runsdc.py
from problem import problem
from sdc import sdc_step
import numpy as np
import copy

M = 5
P = 4
tstart = 0.0
tend   = 10.0
nsteps = 4
dt = (tend - tstart)/float(nsteps)

lambda_1 = -0.1
lambda_2 = -1.0
prob = problem(lambda_1, lambda_2)

K_iter = 12
u_    = np.zeros((M,1,1))
usub_ = np.zeros((M,P,1))
u     = np.zeros((M,1,1))
usub  = np.zeros((M,P,1))
u0    = 2.0
u_ex  = u0*np.exp(tend*(lambda_1+lambda_2))

for n in range(nsteps):
  tstart = float(n)*dt
  tend   = float(n+1)*dt
  sdc    = sdc_step(M, P, tstart, tend, prob)
  
  # reset buffers to zero
  u      = np.zeros((M,1,1))
  usub   = np.zeros((M,P,1))
  u_     = np.zeros((M,1,1))
  usub_  = np.zeros((M,P,1))
  fu     = np.zeros((M, 1))
  fu_sub = np.zeros((M,P,1))

  sdc.predict(u0, u_, usub_, fu, fu_sub)
  for k in range(K_iter):
    sdc.sweep(u0, u, usub, fu, fu_sub)
    u_    = copy.deepcopy(u)
    usub_ = copy.deepcopy(usub)

  print ("+++ step: %3i +++ " % n)
  print ("standard residual: %5.3e " % sdc.residual(u0, u))
  print ("embedded residual: %5.3e " % sdc.sub_residual(u0, usub))
  u0 = u[M-1]

###
print ("error:             %5.3e " % (abs(u0 - u_ex)/abs(u_ex)))

File: multirateSdc/c++/src/runsdc.cc
#include "MRSdc.h"
#include <array>
#include <cmath>

struct Problem
{
	typedef std::array<double, 1u > Vec;
	double l1, l2;
	double fac;

	Problem(double l1, double l2):
		l1(l1), l2(l2)
	{}

	inline void fast(double t, const Vec& in, Vec& out)
	{ out[0] = l2*in[0]; }

	inline void slow(double t, const Vec& in, Vec& out)
	{ out[0] = l1*in[0]; }

	inline void updateMatrix(double t, double a)
	{ fac = 1.0/(1.0-a*l1); }

	inline void solveMaJ(const Vec& in, Vec& out)
	{ out[0] = fac*in[0]; }
};

namespace std
{
	template<typename T, unsigned long s >
	void axpy(double fac, const array<T, s>& x, array<T, s>& y)
	{
		for(unsigned i(0); i < s; ++i)
			y[i] += fac*x[i];
	}

	template<typename T, unsigned long s >
	void setValue(array<T, s>& dest, const T& v)
	{ for(T& d:dest) d= v; }

	template<unsigned long s >
	double norm(const array<double, s >& a)
	{
		double ret(0.0);
		for(const auto& d:a) ret =max(ret, abs(d));
		return ret;
	}

	template<unsigned long s>
	array<double, s > operator-(const array<double, s >& l, const array<double, s >& r)
	{ array<double, s> ret; for(unsigned i(0); i < s ; ++i) ret[i] = l[i]-r[i]; return ret; }
}

using namespace std;
int main(int argc, char* argv[])
{
	double l1(-0.1), l2(-1.0);
	double t0 = 0.0;
	double te   = 10.0;
	unsigned nStep(1);
	Problem::Vec u0({2.0});
	double u_ex  = u0[0]*exp((te-t0)*(l1+l2));

	Problem problem(l1, l2);

	unsigned kIter(12);
	typedef MRSdc<Problem::Vec, 5, 4> Method;
	Method sdc(kIter, "radau_right", "equi_noleft");
	std::cout.precision(8);
#if 0
	double dt=(te-t0)/nStep;
	for(unsigned s(0); s < nStep; ++s) {
		double t0_=t0+s*dt; 
		double te= t0_+dt;
		sdc.predict(problem, u0, t0_, te);
		for(unsigned k(0); k < kIter; ++k) {
			sdc.sweep(problem, u0, t0_, te);
		}
		cout << "standard residual: " << sdc.residual(u0) << std::endl;
		cout << "embedded residual: " << sdc.sub_residual(u0) << std::endl;
		u0[0] = sdc.us[Method::M-1][0];
	}
	cout << "error:" << abs(u0[0]-u_ex)/abs(u_ex) << endl;
#else
	unsigned nTest(10);
	for(unsigned n(0); n < nTest; ++n, nStep *=2) {
		u0[0] = 2.0;
		sdc.solve(problem, u0, t0, te, nStep);
		cout << "error(" << nStep << "):" << abs(u0[0]-u_ex)/abs(u_ex) << endl;
	}
#endif
	return 0;
}
